pkg/controller/podautoscaler/horizontal_test.go

```diff
@@ -4180,3 +4180,102 @@ func TestNoScaleDownOneMetricEmpty(t *testing.T) {
+
+	processed := make(chan string, hpaCount)
+
+	testClient := &fake.Clientset{}
+	testScaleClient := &scalefake.FakeScaleClient{}
+	testMetricsClient := &metricsfake.Clientset{}
+	metricsClient := metrics.NewRESTMetricsClient(
+		testMetricsClient.MetricsV1beta1(),
+		&cmfake.FakeCustomMetricsClient{},
+		&emfake.FakeExternalMetricsClient{},
+	)
+
+	hpaList := [hpaCount]autoscalingv2.HorizontalPodAutoscaler{}
+	var minReplicas int32 = 1
+	for i := 0; i < hpaCount; i++ {
+		h := autoscalingv2.HorizontalPodAutoscaler{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("dummy-hpa-%v", i),
+				Namespace: testNamespace,
+			},
+			Spec: autoscalingv2.HorizontalPodAutoscalerSpec{
+				ScaleTargetRef: autoscalingv2.CrossVersionObjectReference{
+					APIVersion: "apps/v1",
+					Kind:       "Deployment",
+					Name:       fmt.Sprintf("dummy-target-%v", i),
+				},
+				MinReplicas: &minReplicas,
+				MaxReplicas: 10,
+			},
+			Status: autoscalingv2.HorizontalPodAutoscalerStatus{
+				CurrentReplicas: 1,
+				DesiredReplicas: 1,
+				LastScaleTime:   &metav1.Time{Time: time.Now()},
+			},
+		}
+		hpaList[i] = h
+	}
+
+	testClient.AddReactor("list", "horizontalpodautoscalers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		obj := &autoscalingv2.HorizontalPodAutoscalerList{
+			Items: hpaList[:],
+		}
+		return true, obj, nil
+	})
+
+	testClient.AddReactor("update", "horizontalpodautoscalers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		handled, obj, err := func() (handled bool, ret *autoscalingv2.HorizontalPodAutoscaler, err error) {
+			obj := action.(core.UpdateAction).GetObject().(*autoscalingv2.HorizontalPodAutoscaler)
+			assert.Equal(t, testNamespace, obj.Namespace, "the HPA namespace should be as expected")
+
+			return true, obj, nil
+		}()
+
+		processed <- obj.Name
+
+		return handled, obj, err
+	})
+
+	informerFactory := informers.NewSharedInformerFactory(testClient, controller.NoResyncPeriodFunc())
+
+	hpaController := NewHorizontalController(
+		testClient.CoreV1(),
+		testScaleClient,
+		testClient.AutoscalingV2(),
+		testrestmapper.TestOnlyStaticRESTMapper(legacyscheme.Scheme),
+		metricsClient,
+		informerFactory.Autoscaling().V2().HorizontalPodAutoscalers(),
+		informerFactory.Core().V1().Pods(),
+		100*time.Millisecond,
+		5*time.Minute,
+		defaultTestingTolerance,
+		defaultTestingCPUInitializationPeriod,
+		defaultTestingDelayOfInitialReadinessStatus,
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	informerFactory.Start(ctx.Done())
+	go hpaController.Run(ctx, 5)
+
+	timeoutTime := time.After(5 * time.Second)
+	timeout := false
+	processedHPA := make(map[string]bool)
+	for i := 0; i < hpaCount && timeout == false; {
+		select {
+		case hpaName := <-processed:
+			processedHPA[hpaName] = true
+			i++
+		case <-timeoutTime:
+			timeout = true
+		}
+	}
+
+	assert.Equal(t, hpaCount, len(processedHPA), "Expected to process all HPAs")
```

Comment:
You said that you also verify that no concurrent writes happened, how do you (and can you?) verify that?

Reply:
Concurrent access would cause error (panic) in the controller, right? And I suppose that would cause the test failure, or am I mistaken?

Reply:
Honestly I don't know. You could remove your added mutexes temporarily, run the test locally and verify it fails :)

Reply:
Yeah, it fails:
```
fatal error: concurrent map writes

goroutine 44 [running]:
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).recordInitialRecommendation(...)
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:589
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).reconcileAutoscaler(0xc00018a380, {0x27d8238, 0xc0000ac500}, 0xc001082360?, {0xc0005be4a0, 0x1b})
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:636 +0x95d
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).reconcileKey(0xc00018a380, {0x27d8238, 0xc0000ac500}, {0xc0005be4a0, 0x1b})
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:388 +0x131
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).processNextWorkItem(0xc00018a380, {0x27d8238, 0xc0000ac500})
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:232 +0x115
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).worker(0xc00105cde0?, {0x27d8238, 0xc0000ac500})
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:220 +0x45
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
	/Users/zroubali/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:190 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x0?)
	/Users/zroubali/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157 +0x3e
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x27c0ee0, 0xc00105cde0}, 0x1, 0xc00004e540)
	/Users/zroubali/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
	/Users/zroubali/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:135 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext({0x27d8238, 0xc0000ac500}, 0xc000ed2440, 0x0?, 0x0?, 0x0?)
	/Users/zroubali/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:190 +0x99
k8s.io/apimachinery/pkg/util/wait.UntilWithContext({0x27d8238?, 0xc0000ac500?}, 0x0?, 0x0?)
	/Users/zroubali/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:101 +0x2b
created by k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).Run
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:183 +0x1ef

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0004811e0, {0x257f9a6?, 0x119bd2753e6d2?}, 0x264d770)
	/usr/local/opt/go/libexec/src/testing/testing.go:1494 +0x37a
testing.runTests.func1(0xc0004811e0?)
	/usr/local/opt/go/libexec/src/testing/testing.go:1846 +0x6e
testing.tRunner(0xc0004811e0, 0xc000a33cd8)
	/usr/local/opt/go/libexec/src/testing/testing.go:1446 +0x10b
testing.runTests(0xc0003a1900?, {0x3457100, 0x93, 0x93}, {0x2333de0?, 0xc00012b100?, 0x34678e0?})
	/usr/local/opt/go/libexec/src/testing/testing.go:1844 +0x456
testing.(*M).Run(0xc0003a1900)
	/usr/local/opt/go/libexec/src/testing/testing.go:1726 +0x5d9
main.main()
	_testmain.go:339 +0x1aa

goroutine 9 [select]:
k8s.io/kubernetes/pkg/controller/podautoscaler.TestMultipleHPAs(0xc000481380)
	/Users/zroubali/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal_test.go:4271 +0xc25
testing.tRunner(0xc000481380, 0x264d770)
	/usr/local/opt/go/libexec/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/usr/local/opt/go/libexec/src/testing/testing.go:1493 +0x35f
...
```